elk-3 add readmes and documentation (#35)
larry-robotics/elkodon: README.md

performance or modularity, Elkodon is your answer. Welcome to a new era of
inter-process communication.

## Performance

```mermaid
gantt
    title Latency (in ns) - 64b payload
    dateFormat X
    axisFormat %s

    section elkodon
    240 : 0, 240
    section iceoryx
    1000 : 0, 1000
    section MQueue
    700 : 0, 700
    section UDS
    1500 : 0, 1500
```

```mermaid
gantt
    title Latency (in ns) - 64kb payload
    dateFormat X
    axisFormat %s

    section elkodon
    240 : 0, 240
    section iceoryx
    1000 : 0, 1000
    section MQueue
    14000 : 0, 14000
    section UDS
    23000 : 0, 23000
```

**Benchmark-System**

- **CPU:** Intel(R) Core(TM) i7-10875H CPU @ 2.30GHz
- **OS:** Linux 6.5.9-arch2-1 #1 SMP PREEMPT_DYNAMIC GNU/Linux
- **Compiler:**
  - rustc 1.72.1
  - gcc 13.2.1 20230801

## Getting Started

### Publish Subscribe

This minimal example showcases a publisher sending the number 1234 every second,
while a subscriber efficiently receives and prints the data.

**publisher.rs**

```rust
use elkodon::prelude::*;
use elkodon_bb_posix::signal::SignalHandler;

fn main() {
    let service_name = ServiceName::new(b"My/Funk/ServiceName").unwrap();

>     let service = zero_copy::Service::new(&service_name)

Author: Mmmmhhh, in the end it is always IPC but it may makes sense to rename process local into single address space. So if you have a operating system with one big address space, it is the thing for you ... or for process local communication.

So, it may makes sense to rename it into
* `process_local` -> `single_address_space`
* `zero_copy` -> `posix`
* upcoming other variants could `qemu`, `hypervisor` and `vulkan`

Maybe there also comes a copy variant, then I would call it somehow like `posix_non_zero_copy` or so.

@elBoberido What do you think?

>     let service = zero_copy::Service::new(&service_name)

Reviewer: I thinks it moves into the right direction but still does not yet feel completely right. Have to take council with my pillow.

        .publish_subscribe()
        .open_or_create::<usize>()
        .expect("failed to create/open service");

    let publisher = service
        .publisher()
        .create()
        .expect("failed to create publisher");

>     while !SignalHandler::was_ctrl_c_pressed() {

Author: I got inspired by: https://learn.microsoft.com/en-us/windows/console/setconsolectrlhandler and https://crates.io/crates/ctrlc

`Ctrl + c` (`strg` is german for steuerung I think) would emit a `SIGTERM` so whenever this process receives a sigterm this loop would exit.

What about `sigterm_received`?

>     while !SignalHandler::was_ctrl_c_pressed() {

Reviewer: `sigterm` is also quite low level terminology. What about `termination_requested` like you used in the iceoryx implementation?

# elkodon - Zero-Copy Lock-Free IPC Purely Written In Rust

| [benchmarks](benchmarks/README.md) | [best practices](BEST_PRACTICES.md) | [changelog](CHANGELOG.md) | [contributing](CONTRIBUTING.md) | [examples](examples/README.md) | [faq](FAQ.md) | [roadmap](ROADMAP.md) |

 1. [Introduction](#introduction)
 2. [Performance](#performance)
 3. [Getting Started](#getting-started)
    1. [Publish Subscribe](#publish-subscribe)
    2. [Events](#events)
 4. [Supported Platforms](#supported-platforms)
 5. [Language Bindings](#language-bindings)
 6. [Thanks To All Contributors](#thanks-to-all-contributors)

## Introduction

Welcome to Elkodon, the efficient, and ultra-low latency inter-process communication
middleware. This library is designed to provide you with fast and reliable
zero-copy and lock-free inter-process communication mechanisms.

Elkodon is all about providing a seamless experience for inter-process
communication, featuring versatile messaging patterns. Whether you're diving
into publish-subscribe, events, or the promise of upcoming features like
request-response, pipelines, and blackboard, Elkodon has you covered.

> One of the features of Elkodon is its consistency. Regardless of payload size,
> transmission latency remains constant, ensuring a predictable and reliable
> communication experience.

Reviewer: ```suggestion
One of the features of Elkodon is its consistently low transmission latency
regardless of payload size, ensuring a predictable and reliable
communication experience.
```

cargo run --example event_notifier
```

**Terminal 2:**

```sh
cargo run --example event_listener
```

## Supported Platforms

The support levels can be adjusted when required.

| Operating System | Current Support Level | Target Support Level |
|------------------|:---------------------:|---------------------:|
| Android          | tier 3                | tier 1               |
| FreeBSD          | tier 1                | tier 1               |
| FreeRTOS         | tier 3                | tier 2               |
| iOS              | tier 3                | tier 2               |
| Linux (x86_64)   | tier 1                | tier 1               |
| Linux (aarch64)  | tier 1                | tier 1               |
| Linux (32-bit)   | tier 3                | tier 1               |
| Mac OS           | tier 3                | tier 2               |
| QNX              | tier 3                | tier 1               |
| WatchOS          | tier 3                | tier 2               |
| Windows          | tier 2                | tier 2               |

- **tier 1** - All safety and security features are working.
- **tier 2** - Works with a restricted security and safety feature set.

> - **tier 3** - Not yet working but planned.

Reviewer: I think I was a bit imprecise in my last comment. I did not meant to remove the `State` column but to extend the `support` column. I'm fine with keeping the columns as they are now but I would replace some (not all) of the tier 3 with just `planned`. Limux (32-bit) would be tier 3 and QNX just planned. Currently this gives the impression that there is already code written specifically for those platforms but overall it is not yet working.

> | Operating System | State        | Current Support Level | Target Support Level |
> |------------------|:-------------|:---------------------:|---------------------:|
> | Android          | planned      | tier 3                | tier 1               |
> | FreeBSD          | done         | tier 1                | tier 1               |
> | FreeRTOS         | planned      | tier 3                | tier 2               |
> | iOS              | planned      | tier 3                | tier 2               |
> | Linux (x86_64)   | done         | tier 1                | tier 1               |
> | Linux (aarch64)  | done         | tier 1                | tier 1               |
> | Linux (32-bit)   | in-progress  | tier 3                | tier 1               |
> | Mac OS           | in-progress  | tier 3                | tier 2               |
> | QNX              | planned      | tier 3                | tier 1               |
> | WatchOS          | planned      | tier 3                | tier 2               |
> | Windows          | done         | tier 2                | tier 2               |
> 
> - **tier 1** - All safety and security features are working.
> - **tier 2** - Works with a restricted security and safety feature set.
> - **tier 3** - Not yet working but planned.

Reviewer: What do you think of this?

```suggestion
| Operating System | State        | Current Support Level | Target Support Level |
|------------------|:-------------|:---------------------:|---------------------:|
| Android          | planned      | ---                   | tier 1               |
| FreeBSD          | done         | tier 1                | tier 1               |
| FreeRTOS         | planned      | ---                   | tier 2               |
| iOS              | planned      | ---                   | tier 2               |
| Linux (x86_64)   | done         | tier 1                | tier 1               |
| Linux (aarch64)  | done         | tier 1                | tier 1               |
| Linux (32-bit)   | in-progress  | tier 3                | tier 1               |
| Mac OS           | in-progress  | tier 3                | tier 2               |
| QNX              | planned      | ---                   | tier 1               |
| WatchOS          | planned      | ---                   | tier 2               |
| Windows          | done         | tier 2                | tier 2               |

- **tier 1** - All safety and security features are working.
- **tier 2** - Works with a restricted security and safety feature set.
- **tier 3** - Work in progress. Might compile and run or not.
```

I'm even a bit hesitant to mark Linux as done and tier 1 at the moment. Not all communication features are implemented and I don't know the current state of all the implemented safety features.
